### PR TITLE
feat(moe): align unified MoE do_finalize behavior with flat API

### LINKED ISSUE #3926: test(moe): cover do_finalize=False expert_weights dtype in the unified MoE fuzzer
## Summary

The trtllm-gen routing kernel always emits expert weights in **bfloat16** (`routingData.mDtypeOutput` is hard-set to `Bfloat16` for every routing method in `csrc/trtllm_fused_moe_runner.cu`), independent of the `routing_logits` dtype. On the `do_finalize=False` path the `expert_weights` buffer is surfaced to the caller verbatim, so its dtype must be bf16 even when the routing logits are fp32 (the conventional DeepSeekV3 case). A mismatched fp32 label there reinterprets bf16 bytes as fp32 → garbage.

## Coverage gap

The routing-aware unified MoE fuzzer added in #3892 (`tests/moe/test_unified_moe_fuzz.py`) already parametrizes `routing_method × routing_input_mode × logits_dtype`, but it does not assert the dtype of the returned `expert_weights` on the `do_finalize=False` path. That specific combination (fp32 logits + `do_finalize=False`) is currently uncovered.

## Ask

Extend the unified MoE fuzzer (see #3892 for the relevant axes / oracle / dispatch code) so that, for each trtllm-gen fused-MoE operator:

- `trtllm_bf16_moe`
- `trtllm_fp8_per_tensor_scale_moe`
- `trtllm_fp8_block_scale_moe`
- `trtllm_mxint4_block_scale_moe`

run with fp32 (e.g. DeepSeekV3) routing logits and `do_finalize=False`, and assert the returned `expert_weights.dtype == torch.bfloat16` regardless of the `logits_dtype` axis.

This closes the fp32-logits + unfinalized path in the fuzzer and guards against future regressions of the routing-output dtype contract.

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Aligns Unified MoE `do_finalize=False` behavior with the TRTLLM Flat API.
- Returns `[gemm2_output, expert_weights, expanded_idx_to_permuted_idx]`
- Enables unfinalized output for TRTLLM BF16, FP8 block, FP8 per-tensor, FP4, MxInt4, and MXFP8 runners
- Returns BF16 expert weights for `FromLogits` and `PackedPrecomputed` routing
- Preserves caller-provided BF16/FP32 weights for `UnpackedPrecomputed` routing
- Supports BF16 and FP32 routing logits, including MxInt4 logits and bias
- Correctly returns packed FP4 routing weights without accessing an undefined launcher tensor
- Models finalized and unfinalized return contracts in fake implementations
- Adds numerical host-side recombination and permutation validation

## 🔍 Related Issues

Closes #3926

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used my preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests
Tested in `flashinfer/flashinfer-ci-cu130:latest` on SM100:
- 18/18 targeted unfinalized fuzz configurations passed
- Unified MoE contract and runner-support tests passed
- MxInt4 FP32 logits and bias tests passed
- Packed-precomputed tests passed across all targeted TRTLLM variants

## Reviewer Notes
Please focus on:
- Unified and Flat API unfinalized return-contract parity
- Ownership and unpacking of routing-weight buffers
- Packed FP4 expert-weight handling
- FakeTensor output contracts
- MxInt4 FP32 routing logits and bias support

## Summary by CodeRabbit

- **New Features**
  - Added non-finalized Mixture-of-Experts execution across multiple routing and quantization modes.
  - Improved handling of intermediate outputs, expert weights, LoRA outputs, and dynamic output sizes.
  - Expanded support for FP32 routing logits and DeepSeekV3 routing inputs.
  - Improved backend capability reporting when configurations are unsupported.

- **Documentation**
  - Clarified finalization behavior, intermediate output formats, routing-weight types, and backend requirements.

- **Tests**
  - Added coverage for non-finalized outputs, routing configurations, output integrity, and reference accuracy.

### CODE REVIEW

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: flashinfer/fused_moe/api.py [hunk 1]
```diff
@@ -228,8 +228,12 @@ class MoEFinalizeConfig:
     do_finalize : bool
         Whether to apply routing-weight scaling and accumulate the per-expert
         partial results into the output.  ``False`` returns the unreduced
-        per-expert partials, leaving the combine to the caller — only the
-        backends that advertise it support this.
+        TRTLLM intermediates as ``[gemm2_output, expert_weights,
+        expanded_idx_to_permuted_idx]``, leaving the combine to the caller.
+        For FromLogits routing, the routing kernel emits ``expert_weights`` in
+        bfloat16 regardless of the routing-logits dtype. Precomputed routing
+        preserves the caller-provided weights dtype. Only backends that
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/api.py]: `PackedPrecomputed` can't preserve an FP32 dtype for two reasons:

1. **Packing narrows the weights.** `_pack_prerouted_topk_ids` (`runners.py:126-131`) does `topk_weights.to(torch.bfloat16).view(torch.int16)` and folds the low 16 bits into the int32 expert id.
2. **The returned buffer isn't the caller's tensor** — each packed branch passes a separate kernel-side BF16 buffer.

Only `UnpackedPrecomputed` forwards `topk_weights` through unchanged (`runners.py:1471-1472`). Docstring updated to say that.

Chasing this surfaced a real bug: on `PackedPrecomputed` + `do_finalize=False`, `TrtllmFp4RoutedRunner` crashed with `from_dlpack received an invalid capsule`. `FP4BlockScaleLauncher::prepare_routing()` borrows the caller's buffer and never allocates `FusedMoeLauncher::expert_weights`, yet `run()` still assigns that member into the result — so FFI[1] is an undefined Tensor. Fixed by filling the caller-supplied buffer with the routing weights narrowed to BF16.

Separately: the flat `trtllm_fp4_block_scale_moe` op has the same constraint but leaves its buffer **uninitialized** (`core.py:3767-3773`), so it returns garbage `expert_weights` for packed + `do_finalize=False`. The unified runner deliberately diverges to be correct; shape/dtype/ABI unchanged. Filing that separately.

Added six fuzz configs (seeds 900070-900075) covering packed + unfinalized across every TRTLLM variant — the `nvfp4` one caught  18/18 unfinalized, 313 passed + 2skipped (unit), 195 passed + 2 skipped (fuzz).

[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/api.py]: @coderabbitai The docstring fix is on the branch — your check ran against the
pre-rebase SHA. It landed in `c37c3128`, which is now `HEAD` after a rebase onto
upstream ToT and a force-push:

> ``PackedPrecomputed`` routing also yields bfloat16 weights: the caller's values
> are narrowed to bfloat16 when packed into the top-k ids. Only
> ``UnpackedPrecomputed`` routing preserves the caller-provided weights dtype,
> since it forwards ``topk_weights`` to the kernel unchanged.

Your learning note matches the code exactly. Agreed the FP4 buffer fix is a
separate concern — it came out of chasing this one.

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: flashinfer/fused_moe/api.py [hunk 2]
```diff
@@ -228,8 +228,15 @@ class MoEFinalizeConfig:
     do_finalize : bool
         Whether to apply routing-weight scaling and accumulate the per-expert
         partial results into the output.  ``False`` returns the unreduced
-        per-expert partials, leaving the combine to the caller — only the
-        backends that advertise it support this.
+        TRTLLM intermediates as ``[gemm2_output, expert_weights,
```
Comment (Aneureka) [inline flashinfer/fused_moe/api.py:231]: What about the other MoE backends, do they support / follow the same behavior when `do_finalize=False`? Asking this because the comment here seems specific to the TRTLLM MoE backend but we're designing the unified MoE API behavior.

**Review by Aneureka — APPROVED** [review]

File: flashinfer/fused_moe/api.py (hunk 2 above)
Comment (feih-nv) [reply flashinfer/fused_moe/api.py:231]: Only trtllm supports it, `do_finalize` was there in `ExecutionConfig` when we submitted the initial unified API change (pr #3093), and later be moved to `MoEFinalizeConfig` (via pr #4385). 

I agree we are designing it as an unified API behavior that is not limited to trtllm, which needs to be reviewed again perhaps. The good thing is that it is by default `true` so that other backends won't be blocked. Going forward I think we may define it as `optional`, as an alternative option to make things more clear (if we don't think it should be dropped from the api).
